### PR TITLE
Added QNN EP for ARM64 builds.

### PR DESCRIPTION
**Pull request by azchohfi**

### CODE REVIEW

**Comment by azchohfi** [comment]
Not ready for review since the exported projects are not great yet... I'm not sure what to do there...

File: AIDevGallery/Models/ModelCompatibility.cs [hunk 1]
```diff
@@ -23,7 +23,8 @@ public static ModelCompatibility GetModelCompatibility(ModelDetails modelDetails
 
         ModelCompatibilityState compatibility;
         if (modelDetails.HardwareAccelerators.Contains(HardwareAccelerator.CPU) ||
-            (modelDetails.HardwareAccelerators.Contains(HardwareAccelerator.DML) && modelDetails.SupportedOnQualcomm == true))
+            (modelDetails.HardwareAccelerators.Contains(HardwareAccelerator.DML) && modelDetails.SupportedOnQualcomm == true) ||
```
Comment (nmetulev) [inline AIDevGallery/Models/ModelCompatibility.cs]: We should disable all DML models on Arm64

File: AIDevGallery/Samples/SharedCode/EmbeddingGenerator.cs [hunk 2]
```diff
@@ -44,6 +45,15 @@ public EmbeddingGenerator(string modelPath, HardwareAccelerator hardwareAccelera
         {
             _sessionOptions.AppendExecutionProvider_DML(DeviceUtils.GetBestDeviceId());
         }
+        else if (hardwareAccelerator == HardwareAccelerator.QNN)
+        {
+            Dictionary<string, string> options = new()
+            {
+                { "backend_path", "QnnHtp.dll" }
```
Comment (nmetulev) [inline AIDevGallery/Samples/SharedCode/EmbeddingGenerator.cs]: we should include the high performance kv here too

File: AIDevGallery/Samples/ModelsDefinitions/embeddings.modelgroup.json [hunk 3]
```diff
@@ -18,7 +18,8 @@
             "Description": "This is an embedding model executed on the CPU or GPU",
             "HardwareAccelerator": [
               "CPU",
-              "DML"
+              "DML",
+              "QNN"
```
Comment (nmetulev) [inline AIDevGallery/Samples/ModelsDefinitions/embeddings.modelgroup.json]: we need to do some testing here before enabling the models for qnn - I'm seeing both the cpu and npu running which might mean the model is falling back to cpu and it's not optimized for qnn

File: AIDevGallery/Samples/SharedCode/EmbeddingGenerator.cs (hunk 2 above)
Comment (nmetulev) [reply AIDevGallery/Samples/SharedCode/EmbeddingGenerator.cs]: Added

File: AIDevGallery/Models/ModelCompatibility.cs (hunk 1 above)
Comment (nmetulev) [reply AIDevGallery/Models/ModelCompatibility.cs]: disabled

File: AIDevGallery/Samples/ModelsDefinitions/embeddings.modelgroup.json (hunk 3 above)
Comment (nmetulev) [reply AIDevGallery/Samples/ModelsDefinitions/embeddings.modelgroup.json]: removed qnn from embeddings models

**Comment by nmetulev** [comment]
@azchohfi, I made few changes to address my initial comments, please take a look.

In addition, the exported projects still need work to include ORT.QNN instead of ORT.DML. Also, the execution provider in the exported project defaults to CPU even if NPU is selected in the app. Once this is added, I would be good to merge this in

File: AIDevGallery/Samples/SharedCode/EmbeddingGenerator.cs (hunk 2 above)
Comment (azchohfi) [reply AIDevGallery/Samples/SharedCode/EmbeddingGenerator.cs]: You removed QNN from the embedding models, so this if statement should be removed.

Comment (nmetulev) [reply AIDevGallery/Samples/SharedCode/EmbeddingGenerator.cs]: I kept it in there for when we do add QNN - doesn't break anything

File: AIDevGallery/Samples/SharedCode/EmbeddingGenerator.cs [hunk 4]
```diff
@@ -44,6 +45,17 @@ public EmbeddingGenerator(string modelPath, HardwareAccelerator hardwareAccelera
...
+        {
+            Dictionary<string, string> options = new()
+            {
+                { "backend_path", "QnnHtp.dll" },
+                { "htp_performance_mode", "high_performance" },
+                { "htp_graph_finalization_optimization_mode", "3" }
+            };
+            _sessionOptions.AppendExecutionProvider("QNN", options);
+            _chunkSize = 8;
+        }
```
Comment (zateutsch) [inline AIDevGallery/Samples/SharedCode/EmbeddingGenerator.cs:58]: Should be removed if we're not doing NPU for all-Mini

**Review by zateutsch — APPROVED** [review]
Approved with one minor comment.

**Pull request merged**